Implement process creation from data contained in uploaded CSV file without catalog search

solth opened the pull request:
This pull request enhances the mass import to allow creating processes directly from data contained in uploaded CSV files, foregoing the requirement to query catalog search interfaces with record identifiers to import metadata. 

To use this functionality, the first column in the first row of an uploade CSV file has to contain the key of a metadata configured as functional metadata `docType` and the first column of each following row - each representing an individual data record to be transformed into a Kitodo process - should contain the key of a division from the current ruleset ("Art des Inhalts" is configured as `docType` in the ruleset used for the following screenshot). In this case, the label in the header of the first column will contain a special icon to show that a metadata of type `docType` has been found. Note that the "Catalog" selection is deactivated when the first column contains document type information:

<img width="1468" alt="Bildschirmfoto 2025-06-20 um 13 51 25" src="https://github.com/user-attachments/assets/5ab3b9cd-73e7-4ad3-8744-3191b1ec445e" />


The previous functionality of importing metadata from catalog search interfaces still exists, though, and can be triggered by the first row of the first column in the uploaded CSV file containing the key of a metadata set up as a functional metadata of type `recordIdentifier` instead of `docType`. The header of the first column will contain a different symbol to signal a valid functional metadata of type `recordIdentifier` has been found. The catalog selection will only be activated when this is the case:

<img width="1465" alt="Bildschirmfoto 2025-06-20 um 14 08 31" src="https://github.com/user-attachments/assets/3fc16ff6-a265-4ce9-9ef7-413f867c4afd" />



If the first column of the uploaded CSV file contains neither functional metadata of type `docType` nor of type `recordIdentifier`, the header of the first column in the record table will be marked in red to signal a wrong configuration and the "Mass import" button will be deactivated:

<img width="1466" alt="Bildschirmfoto 2025-06-20 um 14 00 27" src="https://github.com/user-attachments/assets/dbac2807-1b9e-4b39-a708-4462b8959cb9" />

A popup dialog will be displayed when opening the mass import to explain the usage described above. This dialog can be hidden for the rest of the session by marking the checkbox in the lower left of the dialog.

<img width="1463" alt="Bildschirmfoto 2025-06-20 um 14 10 57" src="https://github.com/user-attachments/assets/81675c21-7e5d-43c4-97ed-b36757e80de7" />


Fixes #6391

henning-gerhardt commented on Kitodo/src/main/java/org/kitodo/production/forms/LoginForm.java:
```diff
@@ -36,6 +36,27 @@ public class LoginForm implements Serializable {
     private static final String INDEXING_PAGE = "system.jsf?tabIndex=";
     private static final String DESKTOP_VIEW = "desktop.jsf";
     private final SecurityAccessService securityAccessService = ServiceManager.getSecurityAccessService();
+    private boolean hideMassImportExplanationDialog = false;
+
+    /**
+     * Get value of 'hideMassImportExplanationDialog', indicating whether the explanation dialog for the mass import
+     * should be shown during this session or not.
+     *
+     * @return value of 'hideMassImportExplanationDialog'
+     */
+    public boolean isHideMassImportExplanationDialog() {
+        return hideMassImportExplanationDialog;
+    }
+
+    /**
+     * Set value of 'hideMassImportExplanationDialog', indicating whether the explanation dialog for the mass import
+     * should be shown during this session or not.
+     *
+     * @param hideMassImportExplanationDialog whether explanation dialog should be shown or not
+     */
+    public void setHideMassImportExplanationDialog(boolean hideMassImportExplanationDialog) {
+        this.hideMassImportExplanationDialog = hideMassImportExplanationDialog;
+    }
```
Should this not go to the `HelperForm` class? There are - in my opinion - already to many things it this class which does not belong in general to the login itself.

BartChris commented on Kitodo/src/main/resources/messages/messages_en.properties:
```diff
@@ -738,8 +738,13 @@ mappingFile.outputFormat=Metadata output format
 massImport=Mass import
+massImport.explanation=You can either query metadata from a search interface for each record in the datatable below or directly create processes just from the data in the datatable.
+massImport.explanation.recordIdentifier=If you want to import metadata from a search interface, the first column in the table has to contain functional metadata configured as '<b>recordIdentifier</b>'.
+massImport.explanation.docType=Directly converting the data from the table into processes requires the first column to contain function metadata configured as '<b>docType</b>' instead.
```
```suggestion
massImport.explanation.docType=Directly converting the data from the table into processes requires the first column to contain functional metadata configured as '<b>docType</b>' instead.
```

solth replied:
Done

BartChris commented on Kitodo/src/main/java/org/kitodo/production/services/data/MassImportService.java:
```diff
@@ -99,10 +107,17 @@ public List<CsvRecord> parseLines(List<String> lines, String separator) throws I
                      .build()) {
+            int targetNumberOfColumns = -1;
             for (String[] entries : csvReader.readAll()) {
                 if (isSingleEmptyEntry(entries)) {
```
This is already in the old code, but i noticed that lines like 
```
;;;;;;;;;;;;;;;;;;
```
get included in the table as empty lines, maybe we should adapt the check to exclude all empty lines:

```java
 for (String[] entries : csvReader.readAll()) {
                if (isLineCompletelyEmpty(entries)) {
                ..
                }
}

private boolean isLineCompletelyEmpty(String[] entries) {
    return entries.length == 0 || 
           Arrays.stream(entries).allMatch(StringUtils::isBlank);
}
```

solth replied:
You are right. I added the additional check to discard completely empty lines.

BartChris commented on Kitodo/src/main/webapp/pages/massImport.xhtml:
```diff
@@ -39,7 +39,7 @@
         <p:commandButton id="importButton"
                          value="#{msgs.massImport}"
-                         disabled="#{empty MassImportForm.records or not MassImportForm.getRulesetConfigurationForOpacImportComplete()}"
+                         disabled="#{empty MassImportForm.records or not MassImportForm.getRulesetConfigurationForOpacImportComplete() or not MassImportForm.firstColumnContainsRecordIdentifierOrDocType()}"
```
I stumbled over this already, as i - after first put 'ID' in the column - put 'recordIdentifier' in the first column. After reading the description again i got that you have the name of the actual ruleset field which has the functional key applied. Probably just a matter of documentation, but it would be great if there was feedback, why exactly the button is disabled, some message like:

"The metadatum in the first column is not marked as a functional metadata of type `docType` or `recordIdentifier`.

BartChris replied:
nevermind, sorry, it works.

BartChris commented:
Switching to an non existing seperator

![image](https://github.com/user-attachments/assets/1d4fad15-ff45-4f61-af9a-4468fbf92d3b)

shows an error message, but only for the split of a second.

BartChris commented:
The ajax call for checking the progress of the mass import gets triggered even when the import is not actually started. When uploading csv records and clicking on mass import without selecting a catalogue, no import is started but, Kitodo starts polling indefinitely. 

![image](https://github.com/user-attachments/assets/1f9ef855-a740-4de1-bf3b-48b00b13a3ac)

BartChris commented:
When adding a new row and filling in values by hand the whole row gets discarded when "Skip empty colums is selected"-
![image](https://github.com/user-attachments/assets/e35315dd-cb30-4853-93b9-e82abead1a79)

BartChris commented on Kitodo/src/main/java/org/kitodo/production/services/data/MassImportService.java:
```diff
@@ -99,10 +107,17 @@ public List<CsvRecord> parseLines(List<String> lines, String separator) throws I
+                    throw new KitodoCsvImportException(Helper.getTranslation("massImport.separatorCountMismatchEntries",
+                            Pattern.quote(separator)));
+                }
+                targetNumberOfColumns = entries.length;
```
Why is the targetNumberOfColumns determined for every row? Is it not enough to calculate it once - for the first row and then compatre every following row to it?

solth commented:
> Switching to an non existing seperator
> 
> ![image](https://private-user-images.githubusercontent.com/7516499/461002521-1d4fad15-ff45-4f61-af9a-4468fbf92d3b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE4ODIwNDYsIm5iZiI6MTc1MTg4MTc0NiwicGF0aCI6Ii83NTE2NDk5LzQ2MTAwMjUyMS0xZDRmYWQxNS1mZjQ1LTRmNjEtYWY5YS00NDY4ZmJmOTJkM2IucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDdUMDk0OTA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2M4Y2M0YzU2NjFkMDQ0NjA5NGZhMGE1N2RhZWIwOWQ3NTQyOTM3ZjIyZDQ0N2M1NjIwNGI2YTgwM2ZkZDdjMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.YXkialiJStCRLt4y4B6KGVSWLhDFi6TE85_QygMnEyI)
> 
> shows an error message, but only for the split of a second.

This was caused by the import progress being updated every second, which you also observed. I now prevent this by keeping the "Start Mass import" button deactivated until a catalog has been selected (in case the first column contains `recordIdentifier` for opac metadata import.

solth commented on Kitodo/src/main/java/org/kitodo/production/forms/massimport/MassImportForm.java:
```diff
@@ -113,17 +121,75 @@ public void handleFileUpload(FileUploadEvent event) {
+                records = new LinkedList<>();
+                success = false;
+            }
+            if (metadataKeys.size() == 1) {
```
I agree. I do not remember if this code fragment was perhaps some kind of "residue" I used during development and which I forget to remove before opening the pull request, but I think it should be totally valid to upload a CSV file with just one column (if this column contains the required functional metadata of type `recordIdentifier` or `docType`), so I removed this check.

solth commented on Kitodo/src/main/java/org/kitodo/production/forms/massimport/MassImportForm.java:
```diff
@@ -113,17 +121,75 @@ public void handleFileUpload(FileUploadEvent event) {
+                success = false;
+            }
+            if (success && skipEmptyColumns) {
+                List<Integer> skipIndices = ServiceManager.getMassImportService().getColumnSkipIndices(records, metadataKeys.size());
+                if (!skipIndices.isEmpty()) {
+                    metadataKeys = discardMetadataKeysOfEmptyColumns(metadataKeys, skipIndices);
+                    records = discardEmptyColumns(records, skipIndices);
+                }
+            }
```
Yes, I think this would work, but than we need the whole business logic in the backing bean, which I would like to avoid. That's the reason I moved the corresponding methods to the `MassImportService` and therefore couldn't facilitate side effects directly changing values of bean properties in a private method. 

I am unsure if this change would have a real effect on performance and whether we chould forego coding guidelines for it. @BartChris what do you think?

BartChris replied:
You are raising good points. I am unsure about performance. Most of the time people will not import more than 2000 records, which should be very fast process. 
Would it been an option to use a static Service method in the MassImportService which just receives the beans and returns a new list and does not mutate the properties directly?

solth replied:
> Would it been an option to use a static Service method in the MassImportService which just receives the beans and returns a new list and does not mutate the properties directly?

I am not sure this would work, as we need to update not one, but _two_ lists (`records` and `metadataKeys`) with the corresponding indices, so we can't just work with a simple list as return value here. Passing the whole bean (or a clone of it), just to access and modify those two lists and then retrieving the updated lists from the cloned form instance seems a little awkward to me.

BartChris commented:
> > When adding a new row and filling in values by hand the whole new inserted row gets discarded when "Skip empty colums" is selected. The row is also discarded when switching the seperator.
> > ![image](https://private-user-images.githubusercontent.com/7516499/461083685-e35315dd-cb30-4853-93b9-e82abead1a79.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE4ODIwNDYsIm5iZiI6MTc1MTg4MTc0NiwicGF0aCI6Ii83NTE2NDk5LzQ2MTA4MzY4NS1lMzUzMTVkZC1jYjMwLTQ4NTMtOTNiOS1lODJhYmVhZDFhNzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDdUMDk0OTA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmIxMjljNzc4NGFiYTkwMzkxOGM0OThlNjM2ZmYxMGQ2NGFjNDY3ZDMwMDhmYjQyMTBhODg0MjhiZGY5Y2MzNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vB8W3VULHhjtVL31AVR9CiyRnehST6AS4AqW0R7aZMg)
> 
> Yes, this is a known issue - and not restricted to the newly introduced "Skip empty columns" switch. I documented the same problem for changing the the separator character in #6191 before.
> 
> Perhaps I can come up with a fix, but if this turns out to be too complex, I would prefer to deal with that issue in a separate pull request. @BartChris Would you be ok with that?

Yeah,totally fine with that. When thinking about that, i am not even sure if the "Add record" button is a required features -this is a csv import in the end.Especially if it adds much complexity.

solth commented:
> When using the CSV import it is not easy to see from the result list, which entry failed. ![image](https://private-user-images.githubusercontent.com/7516499/461068068-fcfd6b82-9ebe-40d3-b6af-d2b65ff92e60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE5MTM4MDIsIm5iZiI6MTc1MTkxMzUwMiwicGF0aCI6Ii83NTE2NDk5LzQ2MTA2ODA2OC1mY2ZkNmI4Mi05ZWJlLTQwZDMtYjZhZi1kMmI2NWZmOTJlNjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDdUMTgzODIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjdmMzA2MzJmMWNmYzQ4YjY1YzRjNTg0YTMxZGI4NTAzYzFkMTlhNGE0ZDg4NDJjNzMwMTgyZTI1N2QzMWM0YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.HpQaRHMowQd9htAeqVKXdYGj0yORzH80EGzT560wKXk)
> 
> Maybe make a "identifying" column selectable in case of a csv data only import, which could be put as Identifier in the statusmap.

I forgot about this comment, sorry. 

I am not sure adding _another_ new configuration or functional metadata just to be shown in this result list isn't a little oversized of an solution, but perhaps we could include metadata from potentially provided `recordIdentifier` metadata (which can be included when creating processes only from the uploaded CSV data without catalog search, albeit then not in the first column!), if a corresponding column exists. That should be a good indicator to identify records.
@BartChris What do you think?

BartChris commented:
> I am not sure adding _another_ new configuration or functional metadata just to be shown in this result list isn't a little oversized of an solution, but perhaps we could include metadata from potentially provided `recordIdentifier` metadata (which can be included when creating processes only from the uploaded CSV data without catalog search, albeit then not in the first column!), if a corresponding column exists. That should be a good indicator to identify records. @BartChris What do you think?

Sounds very good to me!

solth commented:
> Sounds very good to me!

@BartChris I now incorporated available record identifiers into the result list, which should be more informative now:

Successful imports:
<img width="1016" alt="Bildschirmfoto 2025-07-09 um 14 55 31" src="https://github.com/user-attachments/assets/70d6e61e-3c83-441c-9ab5-dfdfe96020f4" />

Failed imports:
<img width="1015" alt="Bildschirmfoto 2025-07-09 um 14 45 09" src="https://github.com/user-attachments/assets/e12aea70-189f-4913-84d5-c6cb328773ab" />

solth commented:
@BartChris does this solution work for you? One potential small change could be the counter I added to the first column (that you can see in the two screenshots about). As you can see, the counter does not reference the rows in the two datatables "failed imports" and "successful imports", but rather denotes the number of the records in the original datatable of records to be imports (thus those number missing from one result table are contained in the other).

After thinking about this a little more, this might not be as helpful as I intended and perhaps is even a little confusing. Perhaps it would make more sense to just count each result table indidivually, thus resulting with the following tables (identical situation as described above):

Successful imports:
<img width="1019" alt="Bildschirmfoto 2025-07-09 um 16 11 45" src="https://github.com/user-attachments/assets/a08f2d0e-9660-46e1-95a5-9a3e9b0bf3d2" />


Failed imports:
<img width="1019" alt="Bildschirmfoto 2025-07-09 um 16 12 10" src="https://github.com/user-attachments/assets/34b51e8d-88be-4d12-a582-ce05c965027f" />

BartChris commented:
Did your changes make it mandatory to have a key in the pure csv import, which is marked as `recordIdentifer` in the ruleset. 
I get the following exception, when providing only columns with keys which are not a `recordIdentifier`.

<img width="1055" height="408" alt="image" src="https://github.com/user-attachments/assets/17da477f-98e3-4e39-afa3-eab9d1b4b685" />

I think that should still be allowed - the `recordIndentifier` is only a **optional** info to display something in the result list and not should not be mandatory for a csv only import.

solth commented:
> Thanks a lot for pointing this out. This change was not intended. I fixed the import so that `recordIdentifier` is not mandatory anymore during process creation solely from uploaded CSV files.

@BartChris I think the quick solution I just added - just catch the `ConfigException` caused by the missing `recordIdentifier` - my not be optimal since it potentially adds thousand of unnecessary entries to the logs files. I will change this by adding another parameter to `getRecordId` to control whether an exception should be thrown or not, if no record identifier was found.

solth commented:
@BartChris I hope the current solution works for you. Thanks again for your review

BartChris commented:
If there are no records in the successMap or in the failureMap we always get an english message:
`No records found.`, which does not get translated.

The message can be localized by adding a localized string emptyMessage fallback to both tables (Hardcoded in german for demonstration):

 <p:dataTable value="#{MassImportForm.failedImports}"
                                 rowIndexVar="index"
                                 var="importResult"
                                 emptyMessage="Keine Datensätze gefunden">

<p:dataTable value="#{MassImportForm.successfulImports}"
                                 rowIndexVar="index"
                                 var="importResult"
                                 emptyMessage="Keine Datensätze gefunden">

BartChris reviewed (approved):
As far as i can see, it looks very good!
